File: ios/CardScanner/CardScannerViewManager.m
//
//  CardScannerViewManager.m
//  CardScanner
//
//  Created by Howl on 01/06/2022.
//  Copyright © 2022 Facebook. All rights reserved.
//

#import "React/RCTUIManager.h"
#import "React/RCTLog.h"
#import "CardScannerViewManager.h"
#import "CardScannerView.h"

@implementation CardScannerViewManager

RCT_EXPORT_MODULE(CardScannerView);

- (UIView *)view
{
    return [[CardScannerView alloc] initWithFrame:[[UIScreen mainScreen] bounds]];
}

RCT_EXPORT_METHOD(toggleFlash: (nonnull NSNumber *)viewTag)
{
    [self.bridge.uiManager addUIBlock:^(__unused RCTUIManager *uiManager, NSDictionary<NSNumber *, CardScannerView *> *viewRegistry) {
        CardScannerView *view = viewRegistry[viewTag];
        [view toggleFlash];
    }];
}

RCT_EXPORT_METHOD(resetResult: (nonnull NSNumber *)viewTag)
{
    [self.bridge.uiManager addUIBlock:^(__unused RCTUIManager *uiManager, NSDictionary<NSNumber *, CardScannerView *> *viewRegistry) {
        CardScannerView *view = viewRegistry[viewTag];
        [view resetResult];
    }];
}

RCT_EXPORT_VIEW_PROPERTY(frameColor, UIColor);

RCT_EXPORT_VIEW_PROPERTY(onDidScanCard, RCTBubblingEventBlock)

+ (BOOL)requiresMainQueueSetup {
    return YES;
}

@end

File: ios/CardScanner/CardScannerViewManager.swift
//
//  CardScannerViewManager.swift
//  CardScanner
//
//  Created by Howl on 04/06/2022.
//  Copyright © 2022 Facebook. All rights reserved.
//
import Foundation

@objc(CardScannerViewManager)
class CardScannerViewManager: RCTViewManager {

    override func view() -> UIView! {
        return CardScannerView(frame: CGRect.init())
    }
    
    @objc func toggleFlash(_ reactTag: NSNumber) {
        self.bridge!.uiManager.addUIBlock { (_: RCTUIManager?, viewRegistry: [NSNumber: UIView]?) in
            let view: CardScannerView = (viewRegistry![reactTag] as? CardScannerView)!
            view.toggleFlash()
        }
    }
    
    @objc func resetResult(_ reactTag: NSNumber) {
        self.bridge!.uiManager.addUIBlock { (_: RCTUIManager?, viewRegistry: [NSNumber: UIView]?) in
            let view: CardScannerView = (viewRegistry![reactTag] as? CardScannerView)!
            view.resetResult()
        }
    }
    
    @objc func startScanCard(_ reactTag: NSNumber) {
        self.bridge!.uiManager.addUIBlock { (_: RCTUIManager?, viewRegistry: [NSNumber: UIView]?) in
            let view: CardScannerView = (viewRegistry![reactTag] as? CardScannerView)!
            view.startCamera()
        }
    }
    
    @objc func stopScanCard(_ reactTag: NSNumber) {
        self.bridge!.uiManager.addUIBlock { (_: RCTUIManager?, viewRegistry: [NSNumber: UIView]?) in
            let view: CardScannerView = (viewRegistry![reactTag] as? CardScannerView)!
            view.stopCamera()
        }
    }
    
    override static func requiresMainQueueSetup() -> Bool {
        return true
    }
}
